### DISCUSSION TITLE
Logging / Module Caching

### DISCUSSION THREAD

**Discussion by atomdmac** [remix-run/remix#3690] [Q&A] [answered]
I'm attempting to use `pino` to log things from the server-side.  Normally, I'd have a module set up like this:

```typescript
import { pino } from "pino";

console.log('Starting Logger...');

export default pino({
  level: LOG_LEVEL,
  transport: {
    targets: [
      {
        level: LOG_LEVEL,
        target: "pino/file",
        options: {
          destination: `${LOG_DIRECTORY}/my-app.log`,
          append: true,
        },
      },
    ],
  },
});
```

While this does work, I've noticed that I see `Starting Logger` on every call to `logger.debug()`.  It seems that the module is not cached and is spawning a new `logger` instance every time my code runs.

Can someone help me understand what's going on here?  Is module caching disabled in Remix?  Is there a better way to handle logging?

Thanks in advance!

**Comment by kiliman** [comment, marked as answer]
In development, Remix App Server purges the require cache on every request. This is mainly to support `<LiveReload/>`.

You'll want to store your pino object on the `global` object to survive the purge.

```js
import pino from "pino";

declare global {
  var logger: pino.Logger;
}

const LOG_LEVEL = process.env.LOG_LEVEL || "info";
const LOG_DIRECTORY = process.env.LOG_DIRECTORY || "./logs";

function getLogger() {
  console.log("Starting Logger...");

  return pino({
    level: LOG_LEVEL,
    transport: {
      targets: [
        {
          level: LOG_LEVEL,
          target: "pino/file",
          options: {
            destination: `${LOG_DIRECTORY}/my-app.log`,
            append: true,
          },
        },
      ],
    },
  });
}

const logger: pino.Logger =
  process.env.NODE_ENV === "development"
    ? global.logger ?? (global.logger = getLogger())
    : getLogger();

export default logger;
```

**Reply by atomdmac** [reply]
@kiliman Thanks for the explanation!  Won't this just export the locally defined `logger` instead of `global.logger`?

**Reply by kiliman** [reply]
Ok, I edited my example.
